lib/utf8_sanitizer.rb

```diff
@@ -0,0 +1,32 @@
+class Utf8Sanitizer
+  def initialize(app)
+    @app = app
+  end
+
+  def call(env)
+    request = Rack::Request.new(env)
+    values = all_values(request.params)
+
+    return [400, {}, ['Bad request']] if invalid_strings(values)
+
+    @app.call(env)
+  end
+
+  private
+
+  def all_values(hash)
+    hash.values.flat_map { |value| value.is_a?(Hash) ? all_values(value) : [value] }
+  end
+
+  def invalid_strings(values)
+    string_values(values).any? { |string| invalid_string?(string) }
+  end
+
+  def string_values(values)
+    values.select { |value| value.is_a?(String) }
+  end
+
+  def invalid_string?(string)
+    !string.force_encoding('UTF-8').valid_encoding?
```

Comment:
My worry is that someone might use a browser that sends something other than UTF-8. Assuming that Rails converts non-UTF-8 to UTF-8 before handing it to the application, then this isn't a problem, and we only have to worry about Rails taking in content that is advertised as UTF-8 by the client. I'm still looking through Rack/Rails stuff to find a path that shows it looking at the input encoding.

Reply:
There's definitely some risk that users will send non-UTF8 data. I don't think rails does anything to parse the charset out of the request Content-Type header.

But returning an HTTP 400 is at least a little better than the current behavior of exploding and returning HTTP 500. I'm more concerned about future use cases where we want to accept file uploads or other fully binary data. Perhaps we could add some configuration option to the middleware so that if we want to whitelist certain parameters as non-UTF-8 we can do so?

Example realistic use case:

Anyone named Boris is going to be unhappy when `Борис` encoded in Windows-1251 triggers the invalid UTF-8 errors. (`Б` is 0xC1 in Windows-1251, which is an invalid UTF-8 byte.)

Reply:
So:
```ruby
boris = "Борис".encode(Encoding::Windows_1251)
=> "\xC1\xEE\xF0\xE8\xF1"

boris.encoding
=> #<Encoding:Windows-1251>

boris.force_encoding('UTF-8')
=> "\xC1\xEE\xF0\xE8\xF1" # byte sequence hasn't changed - just the encoding

boris.force_encoding('UTF-8').encoding
=> #<Encoding:UTF-8>

boris.force_encoding('UTF-8').valid_encoding?
=> false

boris.encode('UTF-8')
=> "Борис"

boris.encode('UTF-8').valid_encoding?
=> true
```

The best might be to convert all incoming string parameters (except those we want as straight-up binary) to UTF-8 and raise a 400 if the conversion raises any errors.

This is as close as I got to finding something in Rails that might be able to do some of this: https://github.com/rails/rails/blob/b5db73076914e7103466bd76bec785cbcfe88875/actionpack/lib/action_controller/metal/parameter_encoding.rb . It points to a solution that might be a bit overkill for us for now.

Reply:
All of these tests pass, so it looks like Rails handles this well:
```ruby
it 'does not return 400 when email has Cyrillic characters' do
  params = { user: { email: 'Борис@test.ru' } }
  post sign_up_register_path, params: params

  expect(response.status).to eq 302
end

it 'does not return 400 when email is in Windows-1251 encoding' do
  params = { user: { email: 'Борис@test.ru'.force_encoding('Windows-1251') } }
  post sign_up_register_path, params: params

  expect(response.status).to eq 302
end


it 'does not return 400 when email is in Windows-1252 encoding' do
  params = { user: { email: 'Борис@test.ru'.force_encoding('Windows-1252') } }
  post sign_up_register_path, params: params

  expect(response.status).to eq 302
end
```

Reply:
Those tests as written aren't testing actual Windows-1251 text. It should be `'Борис'.encode('Windows-1251')`, not `force_encoding()`.

```ruby
>> 'Борис'.force_encoding('windows-1251').bytes
=> [208, 145, 208, 190, 209, 128, 208, 184, 209, 129]
>> 'Борис'.encode('windows-1251').bytes
=> [193, 238, 240, 232, 241]
```

Reply:
Right, as Jim notes, I don't think Rails tags parameters with the charset specified the browser. In a few quick attempts at testing, the app still assumed everything was UTF-8 even when submitted with a charset specifying Windows-1251.

```ruby
>> binary_boris = "Борис".encode(Encoding::Windows_1251).force_encoding(Encoding::BINARY)
=> "\xC1\xEE\xF0\xE8\xF1"
>> binary_boris.encoding
=> #<Encoding:ASCII-8BIT>
>> binary_boris.encode('utf-8')
Encoding::UndefinedConversionError: "\xC1" from ASCII-8BIT to UTF-8
```

I tested this by submitting a password reset using an email address of `"andrew.brody+\xC1\xEE\xF0\xE8\xF1@gsa.gov"`, which threw an HTTP 500 as expected from the code on master.

Also I feel like Binary Boris is the name of some futurist robot.

Reply:
I don't think that's an accurate test. If you submit `"andrew.brody+\xC1\xEE\xF0\xE8\xF1@gsa.gov"`, the best guess for that String's encoding will be `windows-51` (using the rchardet gem for example: https://github.com/jmhodges/rchardet). You can then convert it to UTF-8 without any errors:
```ruby
windows_andy = "andrew.brody+\xC1\xEE\xF0\xE8\xF1@gsa.gov".encode(Encoding::Windows_1251)
windows_andy.encode('utf-8', 'windows-1251')
=> "andrew.brody+Борис@gsa.gov"
```